scripts/restore_empty_po_entries.py

```
@@ -25,16 +26,29 @@
 I18N_DIR = ROOT / "i18n"
 LOCALES = ("zh_CN", "zh_TW", "en_US", "ja_JP", "ko_KR", "es_ES")
 
+# git ref 名与常见 revision 语法（HEAD~1 / HEAD^ / @{upstream} 等）的
+# 安全字符集；list 形式调用 git show，无 shell 解释，仅拦截明显注入字符
+_SAFE_REF_RE = re.compile(r"^[A-Za-z0-9._/\-~^@{}]+$")
+
+
+def is_safe_ref(ref: str) -> bool:
+    """校验 git ref 是否只含安全字符，避免 CLI 输入逃逸进 git show。"""
```

Ruff (0.16.1)

[warning] Docstring contains ambiguous `，` (FULLWIDTH COMMA). Did you mean `,` (COMMA)?

(RUF002)